src/tools/fuzzing.h

```diff
@@ -315,7 +315,21 @@ class TranslateToFuzzReader {
     }
     SmallVector<Type, 2> options;
     options.push_back(type); // includes itself
-    // TODO (GC): subtyping
+    TODO_SINGLE_COMPOUND(type);
+    switch (type.getBasic()) {
+      case Type::anyref:
+        assert(wasm.features.hasAnyref());
+        if (wasm.features.hasReferenceTypes()) {
+          options.push_back(Type::funcref);
+          options.push_back(Type::externref);
+        }
+        if (wasm.features.hasExceptionHandling()) {
+          options.push_back(Type::exnref);
+        }
```

Comment:
Found that at most places, similar checks are not nested inside `if (hasReferenceTypes)`. Not an urgent problem, but I'd imagine that this is going to lead to obscure issues where a user specifies `--enable-anyref` but not `--enable-reference-types` eventually.

One way to tackle this might be to implicitly enable reference types on `setAnyref(true)` respectively `setExceptionHandling(true)`, and implicitly disable anyref and exception handling on `setReferenceTypes(false)`?

Reply:
Yeah this is a tricky issue. So far we've avoided having any implicit dependencies between features to keep the mental model as simple and predictable as possible. One possibility would be to make it a validation error to have `anyref` enabled without having reference types enabled. That's a little more annoying for users, but on the other hand it keeps the features simple and there should not be real users of the `--enable-anyref` feature anyway.

Reply:
I had the same problem for EH, and I've been doing this kind of things for EH features, to avoid the feature dependences.. https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/tools/fuzzing.h#L2659-L2660

Reply:
Added a feature validation step, but naturally this now breaks the fuzzer because it doesn't know of implied features. Any suggestions?

Reply:
Ended up with https://github.com/WebAssembly/binaryen/pull/3109/commits/fb1a789b0efdc88c6522e6986d34f2e101da77a4 for now

Reply:
I don't think the fuzzer will generate types that does not respect dependences. For example, we [check](https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/tools/fuzzing.h#L2659-L2660) for both `FeatureSet::ExceptionHandling` and `FeatureSet::ReferenceTypes` whenever we try to add `exnref` or other EH instructions (By the way I just realized my PR for generating EH instructions for the fuzzer hasn't landed and is pending for months, but anyway). So I guess the fuzzer error you've seen was just complaining about the feature combination itself, right?

So I'm wondering, why do we need to impose the dependences in validator? Unless giving `--enable-exception-handling` without `--enable-reference-types` itself in the command line of a VM is an error, I don't think we should treat it as an error.

Reply:
> So I guess the fuzzer error you've seen was just complaining about the feature combination itself, right?

Correct, just due to adding the feature validation to the validator and the fuzzer picking random feature opts.

> So I'm wondering, why do we need to impose the dependences in validator?

The initial concern was that at most places we have code like

```
if (hasReferenceTypes) {
  add(funcref)
  add(externref)
}
if (hasExceptionHandling) {
  add(exnref);
}
if (hasAnyref) {
  add(anyref);
}
```

instead of

```
if (hasReferenceTypes) {
  add(funcref)
  add(externref)
  if (hasExceptionHandling) {
    add(exnref);
  }
  if (hasAnyref) {
    add(anyref);
  }
}
```

so it seems possible to yield an `exnref` or `anyref` without reference types being enabled, or to otherwise get false positives, and a validation check lets us keep the code as-is.

Alternatives:
* Refactor all relevant code locations to check for all required features or
* Upon enabling exception handling or anyref, also enable reference types implicitly / upon disabling reference types, also disable exception handling and anyref implicitly

Reply:
From a user's perspective:

* If making missing required features validation errors: "Oh, I see"
* If refactoring all code locations: "Why isn't this option picked up, even though it's specified?"
* If enabling features implicitly: "Why does my binary suddenly utilize reference types, even though it isn't specified?"

From our perspective:

* If making missing required features validation errors: -
* If refactoring all code locations: "Look, a new issue"
* If enabling features implicitly: "Look, a new issue"

Reply:
> The initial concern was that at most places we have code like
> 
> ```
> if (hasReferenceTypes) {
>   add(funcref)
>   add(externref)
> }
> if (hasExceptionHandling) {
>   add(exnref);
> }
> if (hasAnyref) {
>   add(anyref);
> }
> ```
> 
> instead of
> 
> ```
> if (hasReferenceTypes) {
>   add(funcref)
>   add(externref)
>   if (hasExceptionHandling) {
>     add(exnref);
>   }
>   if (hasAnyref) {
>     add(anyref);
>   }
> }
> ```
> 
> so it seems possible to yield an `exnref` or `anyref` without reference types being enabled, or to otherwise get false positives, and a validation check lets us keep the code as-is.

I wonder where we do something like the first? The only place I can think of that _creates_ `exnref` or `anyref` is the fuzzer, and I think it handles them correctly (or you handled them correctly in this PR). Validation or checking wouldn't be as problematic as creating them because they don't bother with `anyref` of `exnref` when they don't exist. I could be mistaken; if so please let me know.

Reply:
Checking again, there is one place where we are doing the former

https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/passes/InstrumentLocals.cpp#L170-L193

and I somehow assumed that this is some sort of convention, even though it probably isn't. Would suggest to move the `hasExceptionHandling` into `hasReferenceTypes` there as well so there can't be instrumented `get_exnref`/`set_exnref` functions taking an `exnref` type if reference types is not enabled, but exception handling is. Validating separately still seems fine to me, though, for clarity.

Reply:
(Haven't checked other locations where just `hasExceptionHandling` is checked, but `hasReferenceTypes` is not)

Reply:
Thanks for finding this. That is a special pass that generates imports, and I agree that moving `hasExceptionHandling` condition into `hasReferenceTypes` is the right call.. Can you add that to this PR if possible?

But I don't think there are many, if at all, passes that generates `anyref` of `exnref` out of thin air. (Except for the fuzzer, but it is not a pass) One function that can generate them and can be called from Binaryen passes is `getLeastUpperBound`, as you pointed out, which you handled correctly in this PR I believe. So I don't think there will be mass refactoring necessary. And I'd rather not implicitly enable dependent features; it can be an unexpected behavior for users.

I just checked what V8 and wabt does for feature dependencies, and actually they are both doing the implication thing, so if you only enable EH that automatically enables reference types too, even if you don't explicitly enable it. Maybe we should consider doing a similar thing in future.